File: matlab/half_wave_rectifier.m
function half_wave_rectifier(Vrms, f, E, R, capacity, opts)
    arguments
        Vrms double = 0.0
        f double = 0.0
        E double = 0.0
        R double = 0.0
        capacity double = 0.0
        opts.Vforward double = 0.0
        opts.Ileak double = 0.0
        opts.Itr double = 0.0
        opts.Itf double = 0.0
        opts.Vtr double = 0.0
        opts.Vtf double = 0.0
        opts.inSOC double = 0.0
        opts.chTime double = 0.0 % hours
        opts.alphaStep double = 0.01
        opts.verbose logical = true
    end

    %% Derived quantities
    Vm = sqrt(2) * Vrms;

    %% Log inputs (optional)
    if opts.verbose
        fprintf("Running FWR charging with:\n");
        fprintf(" Vrms = %.2f V\n f = %.1f Hz\n Battery = %.2f V\n", Vrms, f, E);
        fprintf(" R = %.3f ohm,  Capacity = %.2f Ah\n", R, capacity);
    end

    %% Find firing interval
    [alpha1, beta] = find_range_of_alphas_and_beta_with_v_drop(E, opts.Vforward, Vm);

    %% Alpha sweep
    alphas = make_alpha_array(alpha1, beta, opts.alphaStep);

    %% Find firing interval
    [alpha1_bat_charging, beta_bat_charging] = find_range_of_alphas_and_beta(E, Vm);

    %% Alpha sweep
    alphas_bat_charging = make_alpha_array(alpha1_bat_charging, beta_bat_charging, opts.alphaStep);

    %% Time axis
    x = linspace(0, 2*pi, 10000);

    %% PLOTS / SIMULATION
    if E > 0 && R > 0 && capacity > 0 && Vm > 0
        plot_time_of_charging(alphas_bat_charging, beta_bat_charging, Vm, E, R, capacity, true, 'log');
    end
    
    if E > 0 && R > 0 && capacity > 0 && Vm > 0 && opts.inSOC >= 0 && opts.chTime > 0
        plot_final_soc(alphas_bat_charging, beta_bat_charging, Vm, E, R, opts.inSOC, capacity, opts.chTime, true);
    end

    % if E > 0 && R > 0 && capacity > 0 && Vm > 0 && opts.Ileak && opts.Vforward && opts.Itr && opts.Itf
    %     plot_i_thyristor_with_t(x, alpha1, beta, Vm, E, R, opts.Ileak, opts.Vforward, opts.Itr, opts.Itf, f);
    % end
    % 
    % if E > 0 && R > 0 && capacity > 0 && Vm > 0 && opts.Ileak && opts.Vforward && opts.Vtr && opts.Vtf
    %     plot_Vak_with_t(x, alpha1, beta, Vm, E, opts.Vforward, opts.Vtr, opts.Vtf, f);
    % end
    % 
    % if E > 0 && R > 0 && capacity > 0 && Vm > 0 && opts.Ileak && opts.Vforward && opts.Itr && opts.Itf && opts.Vtr && opts.Vtf
    %     plot_power_loss_with_t(alpha1, beta, x, Vm, E, R, opts.Ileak, opts.Vforward, opts.Itr, opts.Itf, opts.Vtr, opts.Vtf, f);
    % end

    if E > 0 && R > 0 && capacity > 0 && Vm > 0 && opts.Ileak && opts.Vforward && opts.Itr && opts.Itf && opts.Vtr && opts.Vtf
        plot_power_loss_with_alpha(alphas, beta, Vm, E, R, opts.Ileak, opts.Vforward, opts.Itr, opts.Itf, opts.Vtr, opts.Vtf, f);
    end
end

%% --- Utility functions ---------------------------------------------------

function out = norm_angle(theta)
    % Normalize angle to [0, 2*pi)
    out = mod(theta, 2*pi);
end

function [t1, t2] = find_range_of_alphas_and_beta(E, Vm)
    frac = E / Vm;
    if abs(frac) > 1
        error('|E/Vm| must be <= 1. Got %g', frac);
    end
    a = asin(frac);
    t1 = norm_angle(a);
    t2 = norm_angle(pi - a);
end

function [t1, t2] = find_range_of_alphas_and_beta_with_v_drop(E, v_drop, Vm)
    frac = (E + v_drop) / Vm;
    if abs(frac) > 1
        error('|E/Vm| must be <= 1. Got %g', frac);
    end
    a = asin(frac);
    t1 = norm_angle(a);
    t2 = norm_angle(pi - a);
end

function arr = make_alpha_array(a1, a2, step)
    % Create array from a1 to a2 (exclusive) with step, handling wrap-around
    if nargin < 3
        step = 0.01;
    end
    if step <= 0
        error('step must be positive');
    end
    if abs(a1 - a2) < eps
        arr = a1;
        return;
    end
    two_pi = 2*pi;
    if a1 < a2
        % mimic numpy.arange(a1, a2, step)
        arr = a1:step:(a2-step);
        if isempty(arr)
            arr = a1; % fallback minimal array
        end
        return;
    end
    % wrap-around
    part1 = a1:step:(two_pi-step);
    part2 = 0:step:(a2-step);
    if ~isempty(part1) && ~isempty(part2)
        arr = [part1, part2];
    elseif ~isempty(part1)
        arr = part1;
    else
        arr = part2;
    end
end

function y = vs_t(x, Vm)
    y = Vm .* sin(x);
end

function i = i_t(x, alpha, beta, Vm, E, esr)
    % scalar-style function (use arrayfun for vector inputs)
    if (alpha <= x) && (x <= beta)
        i = (vs_t(x, Vm) - E) / esr;
    else
        i = 0.0;
    end
end

function i = i_t_including_leakage_simplified(x, alpha, beta, Vm, E, esr, ileak, v_drop)
    if (alpha <= x) && (x <= beta)
        i = (vs_t(x, Vm) - E - v_drop) / esr;
    else
        i = -ileak;
    end
end

function i = i_t_including_leakage(x, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq)
    % Scalar x expected. For vector x, call with arrayfun.
    omega = 2*pi*freq;
    d_theta_r = omega * Itr;
    d_theta_f = omega * Itf;
    
    naturalI = i_t_including_leakage_simplified(x, alpha, beta, Vm, E, esr, ileak, v_drop);
    
    if x < alpha
        i = -ileak;
        return;
    end
    if (alpha <= x) && (x < alpha + d_theta_r)
        [m,b] = line_from_points(alpha, -ileak, alpha + d_theta_r, i_t_including_leakage_simplified(alpha + d_theta_r, alpha, beta, Vm, E, esr, ileak, v_drop));
        i = m*x + b;
        return;
    end
    if (alpha + d_theta_r <= x) && (x < beta)
        i = naturalI;
        return;
    end
    if (beta <= x) && (x < beta + d_theta_f)
        [m,b] = line_from_points(beta, i_t_including_leakage_simplified(beta, alpha, beta, Vm, E, esr, ileak, v_drop), beta + d_theta_f, -ileak);
        i = m*x + b;
        return;
    end
    i = -ileak;
end

function v = Vak_including_drop_simplified(x, alpha, beta, Vm, E, v_drop)
    if (alpha <= x) && (x <= beta)
        v = v_drop;
    else
        v = Vm * sin(x) - E;
    end
end

function plot_Vak_with_t_simplified(x, alpha, beta, Vm, E, v_drop)
    y = arrayfun(@(xi) Vak_including_drop_simplified(xi, alpha, beta, Vm, E, v_drop), x);
    plot(x, y);
    xlabel('Angle (ωt) rad');
    ylabel('Vak (V)');
    title('Voltage across the load Vak vs Angle (ωt) rad');
    grid on;
    figure;
end

function [m, b] = line_from_points(x1, y1, x2, y2)
    m = (y2 - y1) / (x2 - x1);
    b = y1 - m * x1;
end

function v = Vak_including_drop(x, alpha, beta, Vm, E, v_drop, Vtr, Vtf, freq)
    omega = 2*pi*freq;
    d_theta_r = omega * Vtr;
    d_theta_f = omega * Vtf;
    natural_v = vs_t(x, Vm) - E;
    
    if x < alpha
        v = natural_v;
        return;
    end
    if (alpha <= x) && (x < alpha + d_theta_f)
        [m,b] = line_from_points(alpha, vs_t(alpha, Vm)-E, alpha + d_theta_f, v_drop);
        v = m*x + b;
        return;
    end
    if (alpha + d_theta_f <= x) && (x < beta)
        v = v_drop;
        return;
    end
    if (beta <= x) && (x < beta + d_theta_r)
        [m,b] = line_from_points(beta, v_drop, beta + d_theta_r, vs_t(beta + d_theta_r, Vm)-E);
        v = m*x + b;
        return; 
    end
    v = natural_v;
end

function plot_Vak_with_t(x, alpha, beta, Vm, E, v_drop, Vtr, Vtf, freq)
    y = arrayfun(@(xi) Vak_including_drop(xi, alpha, beta, Vm, E, v_drop, Vtr, Vtf, freq), x);
    plot(x, y);
    xlabel('Angle (ωt) rad');
    ylabel('Vak (V)');
    title('Voltage across the load Vak vs Angle (ωt) rad');
    grid on;
    figure; 
end

function plot_i_thyristor_with_t(x, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq)
    y = arrayfun(@(xi) i_t_including_leakage(xi, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq), x);
    plot(x, y);
    xlabel('Angle (ωt) rad');
    ylabel('Current i (A)');
    title('Current i vs Angle (ωt) rad');
    grid on;
    figure;
end

function Iavg = compute_Iavg(Vm, E, esr, alpha, beta)
    integrand = @(xx) arrayfun(@(xval) i_t(xval, alpha, beta, Vm, E, esr), xx);
    result = integral(integrand, 0, 2*pi, 'ArrayValued', true, 'RelTol',1e-6,'AbsTol',1e-9);
    Iavg = (1.0 / (2.0 * pi)) * result;
end

function Irms = compute_i_t_including_leakage_rms(alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq)
    integrand = @(xx) arrayfun(@(xval) i_t_including_leakage(xval, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq).^2, xx);
    integral_val = integral(integrand, 0, 2.0*pi, 'ArrayValued', true, 'RelTol',1e-6,'AbsTol',1e-9);
    Irms = sqrt(integral_val / (2.0 * pi));
end

function P = compute_power_loss(x, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq)
    Vak = Vak_including_drop(x, alpha, beta, Vm, E, v_drop, Vtr, Vtf, freq);
    iVal = i_t_including_leakage(x, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq);
    P = Vak .* iVal + esr .* (iVal.^2);
end

function P_loss_avg = compute_average_power_loss(alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq)
    power_loss_func = @(xx) arrayfun(@(xval) compute_power_loss(xval, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq), xx);
    result = integral(power_loss_func, 0, 2*pi, 'ArrayValued', true, 'RelTol',1e-6,'AbsTol',1e-9);
    P_loss_avg = (1.0 / (2.0 * pi)) * result;
end

function plot_power_loss_with_alpha(alpha_array, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq)
    y = zeros(size(alpha_array));
    for k = 1:numel(alpha_array)
        alpha = alpha_array(k);
        % Irms computed but unused (kept like original)
        Irms = compute_i_t_including_leakage_rms(alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, freq);
        P_loss_avg = compute_average_power_loss(alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq);
        y(k) = P_loss_avg;
    end
    plot(alpha_array, y);
    xlabel('Firing Angle (rad)');
    ylabel('Average Power Loss (W)');
    title('Average Power Loss vs Firing Angle');
    grid on;
    figure;
end

function plot_power_loss_with_t(alpha, beta, x, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq)
    y = arrayfun(@(xi) compute_power_loss(xi, alpha, beta, Vm, E, esr, ileak, v_drop, Itr, Itf, Vtr, Vtf, freq), x);
    plot(x, y);
    xlabel('Angle (ωt) rad');
    ylabel('Power Loss (W)');
    title('Power Loss vs Angle');
    grid on;
    figure;
end

function charging_time = compute_charging_time(Iavg, C)
    % Compute charging time (hours) from capacity C (Ah) and Iavg (A)
    charging_time = nan(size(Iavg));
    if ~isempty(Iavg)
        positive = Iavg > 0;
        if ~isempty(C)
            charging_time(positive) = C ./ Iavg(positive);
        end
    end
end

function plot_time_of_charging(alphas, beta, Vm, E, esr, C, degreesFlag, charging_time_scale)
    if nargin < 7
        degreesFlag = true;
    end
    if nargin < 8
        charging_time_scale = 'linear';
    end
    if degreesFlag
        x = rad2deg(alphas);
        xlabelTxt = 'Firing angle (deg)';
    else
        x = alphas;
        xlabelTxt = 'Firing angle (rad)';
    end
    [Iavg, charging_time] = time_of_charging(Vm, E, esr, C, alphas, beta);
    
    figure('Units','normalized','Position',[0.1 0.1 0.6 0.6]);
    yyaxis left
    plot(x, Iavg, '-o', 'MarkerSize', 4);
    ylabel('Iavg (A)');
    xlabel(xlabelTxt, 'FontSize', 11);
    grid on;
    set(gca,'FontSize',11);
    
    if ~all(isnan(charging_time))
        yyaxis right
        plot(x, charging_time, '-s', 'MarkerSize', 4);
        ylabelText = 'Charging time (h)';
        if exist('charging_time_scale','var') && ~strcmp(charging_time_scale,'linear')
            set(gca, 'YScale', charging_time_scale);
            ylabelText = sprintf('%s (%s)', ylabelText, charging_time_scale);
        end
        ylabel(ylabelText);
    end
    
    title(sprintf('Iavg and charging time vs firing angle (Vrms=%g V, E=%g V, esr=%g ohms, C=%g Ah)', Vm/sqrt(2), E, esr, C));
    figure;
end

function new_soc = compute_final_soc(Iavg, initial_soc, C, required_charging_time)
    % initial_soc in percent (0-100)
    new_soc = (Iavg .* required_charging_time + (initial_soc/100.0) * C) ./ C;
    new_soc = new_soc * 100.0;
    new_soc = min(max(new_soc, 0.0), 100.0);
end

function plot_final_soc(alphas, beta, Vm, E, esr, initial_soc, C, required_charging_time, degreesFlag)
    if nargin < 9
        degreesFlag = true;
    end
    if degreesFlag
        x = rad2deg(alphas);
        xlabelTxt = 'Firing angle (deg)';
    else
        x = alphas;
        xlabelTxt = 'Firing angle (rad)';
    end
    
    [Iavg, ~] = time_of_charging(Vm, E, esr, C, alphas, beta);
    new_soc = compute_final_soc(Iavg, initial_soc, C, required_charging_time);
    
    figure;
    plot(x, new_soc, '-o', 'MarkerSize', 4);
    xlabel(xlabelTxt, 'FontSize', 11);
    ylabel('Final State of Charge (%)', 'FontSize', 11);
    grid on;
    title(sprintf('Final SOC vs Firing Angle\nInit SOC=%g%%, Time=%gh, C=%gAh', initial_soc, required_charging_time, C));
    ylim([0 100]);
    figure;
end

function [Iavg, charging_time] = time_of_charging(Vm, E, esr, C, alphas, beta)
    Iavg = zeros(size(alphas));
    for k = 1:numel(alphas)
        Iavg(k) = compute_Iavg(Vm, E, esr, alphas(k), beta);
    end
    charging_time = compute_charging_time(Iavg, C);
end

File: python/half_wave_rectifier.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import quad

def _norm_angle(theta: float) -> float:
    """Normalize angle to [0, 2*pi).
    Works for scalars and numpy arrays (uses modulo).
    """
    return np.mod(theta, 2 * np.pi)


def find_range_of_alphas_and_beta(E: float, Vm: float):
    """
    Return two angles (alpha1, alpha2) in radians in [0, 2*pi)
    that satisfy sin(theta) = E/Vm. Raises ValueError if |E/Vm|>1.
    The two solutions are arcsin(frac) and pi - arcsin(frac),
    normalized into [0, 2*pi).
    """
    frac = E / Vm
    if abs(frac) > 1.0:
        raise ValueError(f"|E/Vm| must be <= 1. Got {frac}")
    a = np.arcsin(frac)
    t1 = _norm_angle(a)
    t2 = _norm_angle(np.pi - a)
    return t1, t2


def _make_alpha_array(a1: float, a2: float, step: float = 0.01) -> np.ndarray:
    """Create a numpy array of angles from a1 to a2 (exclusive) with step.
    Handles wrap-around if a1 > a2 by concatenating [a1, 2pi) and [0, a2).
    """
    a1 = float(a1)
    a2 = float(a2)
    step = float(step)
    if step <= 0:
        raise ValueError("step must be positive")
    two_pi = 2 * np.pi
    if np.isclose(a1, a2):
        return np.array([a1])
    if a1 < a2:
        return np.arange(a1, a2, step, dtype=float)
    # wrap-around
    part1 = np.arange(a1, two_pi, step, dtype=float)
    part2 = np.arange(0.0, a2, step, dtype=float)
    if part1.size and part2.size:
        return np.concatenate((part1, part2))
    return part1 if part1.size else part2

def vs_t(x, Vm):
    return Vm*np.sin(x)

def i_t(x, alpha, beta, Vm, E, esr):
    if alpha <= x <= beta:
        return (Vm*np.sin(x)-E)/esr
    else:
        return 0.0  # for 2pi-(beta-alpha)
    
def i_t_including_leakage_simplified(x, alpha, beta, Vm, E, esr, ileak):
    if alpha <= x <= beta:
        return (Vm*np.sin(x)-E)/esr
    else:
        return ileak  # for 2pi-(beta-alpha)

def i_t_including_leakage(x, alpha, beta, Vm, E, esr, ileak, tr, tf, freq):
    # Convert rise/fall times to angular ranges
    omega = 2*np.pi*freq
    d_theta_r = omega * tr   # rise-time converted to angle
    d_theta_f = omega * tf   # fall-time converted to angle

    natural_I = i_t(x, alpha, beta, Vm, E, esr)

    # Region 1: before α (Ileak)
    if x < alpha:
        return ileak

    # Region 2: α → α + d_theta_r (rising current to natural I)
    if alpha <= x < alpha + d_theta_r:
        m, b = line_from_points(alpha, ileak, alpha + d_theta_r, i_t(alpha + d_theta_r, alpha, beta, Vm, E, esr))
        return m*x + b  # RISE: ileak → natural

    # Region 3: The natural current region
    if alpha + d_theta_r <= x < beta:
        return natural_I

    # Region 4: β → β + theta_f (falling current back to ileak)
    if beta <= x < beta + d_theta_f:
        m, b = line_from_points(beta, i_t(beta, alpha, beta, Vm, E, esr), beta + d_theta_f, -ileak)
        return m*x + b # FALL: natural → -ileak

    # Region 5: after β + theta_f (-Ileak)
    return -ileak


def Vak_including_drop_simplified(x, alpha, beta, Vm, E, v_drop):
    if alpha <= x <= beta:
        return v_drop
    else:
        return Vm*np.sin(x)-E  # for 2pi-(beta-alpha)

def plot_Vak_with_t_simplified(x, alpha, beta, Vm, E, v_drop):
    y = [Vak_including_drop_simplified(xi, alpha, beta, Vm, E, v_drop) for xi in x]
    plt.plot(x, y)
    plt.xlabel("Angle (rad)")
    plt.ylabel("Vak (V)")
    plt.title("Voltage across the load Vak vs Angle")
    plt.grid(True)
    plt.show()

def line_from_points(x1, y1, x2, y2):
    # Slope (m)
    m = (y2 - y1) / (x2 - x1)
    # Intercept (b)
    b = y1 - m * x1
    return m, b

def Vak_including_drop(x, alpha, beta, Vm, E, v_drop, tr, tf, freq):
    # Convert rise/fall times to angular ranges
    omega = 2*np.pi*freq
    d_theta_r = omega * tr   # rise-time converted to angle 
    d_theta_f = omega * tf   # fall-time converted to angle 

    natural_v = vs_t(x, Vm) - E

    # Region 1: before α (natural waveform)
    if x < alpha:
        return natural_v

    # Region 2: α → α + d_theta_f (falling voltage to v_drop)
    if alpha <= x < alpha + d_theta_f:
        m, b = line_from_points(alpha, vs_t(alpha, Vm)-E, alpha + d_theta_f, v_drop)
        return m*x + b  # FALL: natural → v_drop

    # Region 3: constant v_drop
    if alpha + d_theta_f <= x < beta:
        return v_drop

    # Region 4: β → β + theta_r (rising voltage back to natural)
    if beta <= x < beta + d_theta_r:
        m, b = line_from_points(beta, v_drop, beta + d_theta_r, vs_t(beta + d_theta_r, Vm)-E)
        return m*x + b # RISE: v_drop → natural

    # Region 5: after β + theta_r (natural waveform)
    return natural_v


def plot_Vak_with_t(x, alpha, beta, Vm, E, v_drop, tr, tf, freq):
    y = [Vak_including_drop(xi, alpha, beta, Vm, E, v_drop, tr, tf, freq) for xi in x]
    plt.plot(x, y)
    plt.xlabel("Angle (rad)")
    plt.ylabel("Vak (V)")
    plt.title("Voltage across the load Vak vs Angle")
    plt.grid(True)  
    plt.show()

def plot_i_with_t(x, alpha, beta, Vm, E, esr, ileak, tr, tf, freq):
    y = [i_t_including_leakage(xi, alpha, beta, Vm, E, esr, ileak, tr, tf, freq) for xi in x]
    plt.plot(x, y)
    plt.xlabel("Angle (rad)")
    plt.ylabel("Current i (A)")
    plt.title("Current i vs Angle")
    plt.grid(True)
    plt.show()

def compute_Iavg(Vm, E, esr, alpha, beta) -> float:
    i = lambda x: i_t(x, alpha, beta, Vm, E, esr)
    result, _ = quad(i, 0, 2*np.pi, limit=1000)
    Iavg = (1.0 / (2.0 * np.pi)) * result
    return Iavg

def compute_i_t_including_leakage_rms(alpha, beta, Vm, E, esr, ileak, tr, tf, freq):
    integral, _ = quad(lambda x: i_t_including_leakage(x, alpha, beta, Vm, E, esr, ileak, tr, tf, freq)**2, 0, 2.0 * np.pi, limit=1000)
    return np.sqrt(integral / (2.0 * np.pi))

def compute_power_loss(x, alpha, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq, Irms):
    Vak = Vak_including_drop(x, alpha, beta, Vm, E, v_drop, tr, tf, freq)
    i = i_t_including_leakage(x, alpha, beta, Vm, E, esr, ileak, tr, tf, freq)
    return Vak * i + esr * i**2

def compute_average_power_loss(alpha, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq, Irms):
    power_loss_func = lambda x: compute_power_loss(x, alpha, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq, Irms)
    result, _ = quad(power_loss_func, 0, 2*np.pi, limit=1000)
    P_loss_avg = (1.0 / (2.0 * np.pi)) * result
    return P_loss_avg

def plot_power_loss_with_alpha(alpha_array, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq):
    y = []
    for alpha in alpha_array:
        Irms = compute_i_t_including_leakage_rms(alpha, beta, Vm, E, esr, ileak, tr, tf, freq)
        P_loss_avg = compute_average_power_loss(alpha, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq, Irms)
        y.append(P_loss_avg)
    plt.plot(alpha_array, y)
    plt.xlabel("Firing Angle (rad)")
    plt.ylabel("Average Power Loss (W)")
    plt.title("Average Power Loss vs Firing Angle")
    plt.grid(True)
    plt.show()

def plot_power_loss_with_t(x, Vm, E, esr, ileak, tr, tf, freq, alpha, beta):
    y = [compute_power_loss(xi, alpha, beta, Vm, E, esr, ileak, tr, tf, freq) for xi in x]
    plt.plot(x, y)
    plt.xlabel("Angle (rad)")
    plt.ylabel("Power Loss (W)")
    plt.title("Power Loss vs Angle")
    plt.grid(True)
    plt.show()


def time_of_charging(Vrms, E, esr, C, step: float = 0.01, degrees: bool = True, show: bool = True,
                     charging_time_scale: str = 'linear', initial_soc: float = 0.0, required_charging_time: float = 0.0):
    Vm = np.sqrt(2.0) * Vrms
    alpha1, beta = find_range_of_alphas_and_beta(E, Vm)
    alphas = _make_alpha_array(alpha1, beta, step=step)
    
    # here we compute Iavg for each alpha
    Iavg = []
    for alpha in alphas:
        Iavg.append(compute_Iavg(Vm, E, esr, alpha, beta))
    Iavg = np.array(Iavg)

    # Compute charging time with respect to alphas (C / Iavg) when capacity C provided.
    # Only valid when Iavg > 0. Mask non-positive currents to NaN so they are not plotted.
    charging_time = C/Iavg
    if C is not None:
        # Broadcast: C scalar divided by Iavg array
        charging_time = np.full_like(Iavg, np.nan, dtype=float)
        positive = Iavg > 0
        charging_time[positive] = C / Iavg[positive]

    # Plot
    x = np.degrees(alphas) if degrees else alphas
    xlabel = "Firing angle (deg)" if degrees else "Firing angle (rad)"
    # Larger figure size so small numbers are visible
    fig, ax1 = plt.subplots(figsize=(14, 8))
    ax1.plot(x, Iavg, '-o', markersize=4, label='Iavg (A)', color='tab:blue')
    ax1.set_xlabel(xlabel, fontsize=12)
    ax1.set_ylabel('Iavg (A)', color='tab:blue', fontsize=12)
    ax1.tick_params(axis='y', labelcolor='tab:blue', labelsize=11)
    ax1.tick_params(axis='x', labelsize=11)
    ax1.grid(True, which='both', axis='both', linestyle='--', alpha=0.4)
    if charging_time is not None:
        ax2 = ax1.twinx()
        ax2.plot(x, charging_time, '-s', markersize=4, label='Charging time (s)', color='tab:orange')
        ylabel = 'Charging time (s)'
        if charging_time_scale and charging_time_scale != 'linear':
            ylabel += f' ({charging_time_scale})'
            ax2.set_yscale(charging_time_scale)
        ax2.set_ylabel(ylabel, color='tab:orange', fontsize=12)
        ax2.tick_params(axis='y', labelcolor='tab:orange', labelsize=11)
        # Auto-scale y-axis to fit all data with small padding
        ax2.margins(y=0.1)
    plt.title(f'Iavg and charging time vs firing angle (Vrms={Vrms} V, E={E} V, esr={esr} ohms, C={C} Ah)', fontsize=13, fontweight='bold')
    fig.tight_layout()

    if required_charging_time > 0.0 and initial_soc > 0.0:
        new_soc = (Iavg * required_charging_time  + (initial_soc/100.0)*C) / C
        new_soc = new_soc * 100.0  # convert to percentage
        new_soc = np.clip(new_soc, 0.0, 100.0)   # clip to 100%

        # Create second plot for final state of charge
        fig2, ax3 = plt.subplots(figsize=(14, 6))
        ax3.plot(x, new_soc, '-o', markersize=4, color='tab:green')
        ax3.set_xlabel(xlabel, fontsize=12)
        ax3.set_ylabel("Final State of Charge (%)", fontsize=12, color='tab:green')
        ax3.tick_params(axis='y', labelcolor='tab:green', labelsize=11)
        ax3.tick_params(axis='x', labelsize=11)
        ax3.grid(True, linestyle='--', alpha=0.4)
        plt.title(f"Final SOC vs Firing Angle\nInit SOC={initial_soc}%, Time={required_charging_time}h, C={C}Ah",fontsize=13,fontweight='bold')

        fig2.tight_layout()

    if show:
        plt.show()

    return alphas, Iavg, charging_time    

if __name__ == "__main__":
    Vrms = 60
    E = 12
    esr = 4.256
    capacity = 100
    v_drop = 1.2
    tr = 10*10**(-6)  # rise time in seconds
    tf = 5*10**(-6)  # fall time in seconds
    freq = 50  # frequency in Hz
    ileak = 0.02  # leakage current in Amperes
    Vm = np.sqrt(2)*Vrms
    alpha1, beta = find_range_of_alphas_and_beta(E, Vm)
    alphas = _make_alpha_array(alpha1, beta, step=0.01)
    x = np.linspace(0, 2*np.pi, 10000)
    # plot_Vak_with_t_simplified(x, alpha1, beta, Vm, E, v_drop)
    time_of_charging(Vrms, E, esr, capacity, step=0.01, charging_time_scale='log', initial_soc=30.0, required_charging_time=0.5)
    # plot_Vak_with_t(x, alpha1, beta, Vm, E, v_drop, tr, tf, freq)
    # plot_i_with_t(x, alpha1, beta, Vm, E, esr, ileak=ileak, tr=tr, tf=tf, freq=freq)
    # plot_power_loss_with_t(x, Vm, E, esr, ileak, tr, tf, freq, alpha1, beta)
    # plot_power_loss_with_alpha(alphas, beta, Vm, E, esr, ileak, v_drop, tr, tf, freq)
    # print(compute_i_t_including_leakage_rms(alpha1, beta, Vm, E, esr, ileak, tr, tf, freq))